File: preset/src/main/java/org/openbase/bco/app/preset/PartyLightTileFollowerApp.java
package org.openbase.bco.app.preset;

/*
 * #%L
 * BCO App Preset
 * %%
 * Copyright (C) 2018 - 2020 openbase.org
 * %%
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public
 * License along with this program.  If not, see
 * <http://www.gnu.org/licenses/gpl-3.0.html>.
 * #L%
 */

import org.openbase.bco.dal.lib.layer.unit.Unit;
import org.openbase.bco.dal.remote.action.RemoteAction;
import org.openbase.bco.dal.remote.layer.unit.Units;
import org.openbase.bco.dal.remote.layer.unit.location.LocationRemote;
import org.openbase.bco.dal.control.layer.unit.app.AbstractAppController;
import org.openbase.bco.registry.remote.Registries;
import org.openbase.jul.exception.CouldNotPerformException;
import org.openbase.jul.exception.ExceptionProcessor;
import org.openbase.jul.exception.InstantiationException;
import org.openbase.jul.exception.printer.ExceptionPrinter;
import org.openbase.jul.schedule.SyncObject;
import org.openbase.type.domotic.action.ActionDescriptionType.ActionDescription;
import org.openbase.type.domotic.state.ActivationStateType.ActivationState;
import org.openbase.type.domotic.unit.UnitConfigType.UnitConfig;
import org.openbase.type.domotic.unit.UnitTemplateType.UnitTemplate.UnitType;
import org.openbase.type.domotic.unit.location.LocationConfigType;
import org.openbase.type.vision.HSBColorType.HSBColor;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;
import java.util.concurrent.*;

/**
 * UnitConfig
 */
public class PartyLightTileFollowerApp extends AbstractAppController {

    private Map<Unit, RemoteAction> actionLocationMap;
    private Map<String, LocationRemote> locationRemoteMap;

    private SyncObject taskLock = new SyncObject("TaskLock");

    public PartyLightTileFollowerApp() throws InstantiationException, InterruptedException {
        try {
            Registries.waitForData();
            this.actionLocationMap = new HashMap<>();
            this.locationRemoteMap = new HashMap<>();
        } catch (CouldNotPerformException ex) {
            throw new InstantiationException(this, ex);
        }
    }

    @Override
    public void shutdown() {
        locationRemoteMap.clear();
        super.shutdown();
    }

    private double brightness = 0.50d;
    private HSBColor[] colors = {
            HSBColor.newBuilder().setHue(0d).setSaturation(1.0d).setBrightness(brightness).build(),
            HSBColor.newBuilder().setHue(290d).setSaturation(1.0d).setBrightness(brightness).build(),
            HSBColor.newBuilder().setHue(30d).setSaturation(1.0d).setBrightness(brightness).build(),};

    @Override
    protected ActionDescription execute(final ActivationState activationState) throws CouldNotPerformException, InterruptedException {
        logger.debug("Execute PartyLightTileFollowerApp[" + getLabel() + "]");

        // init tile remotes
        locationRemoteMap.clear();
        for (final UnitConfig locationUnitConfig : Registries.getUnitRegistry(true).getUnitConfigsByUnitType(UnitType.LOCATION)) {
            if (!locationUnitConfig.getLocationConfig().getLocationType().equals(LocationConfigType.LocationConfig.LocationType.TILE)) {
                continue;
            }
            locationRemoteMap.put(locationUnitConfig.getId(), Units.getUnit(locationUnitConfig, false, Units.LOCATION));
        }

        new TileFollower().call();
        return activationState.getResponsibleAction();
    }

    @Override
    protected void stop(final ActivationState activationState) throws InterruptedException, CouldNotPerformException {
        final ArrayList<Future<ActionDescription>> cancelTaskList = new ArrayList<>();
        for (Entry<Unit, RemoteAction> unitActionEntry : actionLocationMap.entrySet()) {
            cancelTaskList.add(unitActionEntry.getValue().cancel());
        }

        for (Future<ActionDescription> cancelTask : cancelTaskList) {
            try {
                cancelTask.get(10, TimeUnit.SECONDS);
            } catch (ExecutionException | TimeoutException ex) {
                ExceptionPrinter.printHistory("Could not cancel action!", ex, logger);
            }
        }
        super.stop(activationState);
    }

    public class TileFollower implements Callable<Void> {

        private final List<String> processedLocations = new ArrayList<>();

        @Override
        public Void call() throws CouldNotPerformException, InterruptedException {
            if (locationRemoteMap.isEmpty()) {
                throw new CouldNotPerformException("No Locations found!");
            }

            LocationRemote locationRemote;

            int colorIndex = 0;
            while (!Thread.currentThread().isInterrupted()) {
                Thread.yield();
                try {
                    // apply updates for next iteration
                    colorIndex = ++colorIndex % colors.length;
                    processedLocations.clear();

                    // select initial room
                    locationRemote = locationRemoteMap.get(getConfig().getPlacementConfig().getLocationId());

                    processRoom(locationRemote, colors[colorIndex]);

                } catch (CouldNotPerformException ex) {
                    ExceptionPrinter.printHistory(new CouldNotPerformException("Skip animation run!", ExceptionProcessor.interruptOnShutdown(ex)), logger);
                }
            }
            return null;
        }

        private void processRoom(final LocationRemote locationRemote, final HSBColor color) throws CouldNotPerformException, InterruptedException {
            logger.debug("Set " + locationRemote + " to " + color + "...");
            try {

                // skip if no colorable light is present
                if (!Registries.getUnitRegistry().getUnitConfigsByLocationIdAndUnitType(locationRemote.getId(), UnitType.COLORABLE_LIGHT).isEmpty()) {
                    if (locationRemote.isConnected() && locationRemote.isDataAvailable()) {
                        final RemoteAction remoteAction = observe(locationRemote.setColor(color, getDefaultActionParameter()));
                        actionLocationMap.put(locationRemote, remoteAction);
                        remoteAction.waitForRegistration();
                    }
                    Thread.sleep(1000);
                }

                // mark as processed
                processedLocations.add(locationRemote.getId());

                // process neighbors
                LocationRemote neighborRemote;
                for (UnitConfig neighborConfig : Registries.getUnitRegistry().getNeighborLocationsByLocationId(locationRemote.getId())) {
                    // skip if already processed
                    if (processedLocations.contains(neighborConfig.getId())) {
                        continue;
                    }

                    neighborRemote = locationRemoteMap.get(neighborConfig.getId());

                    // process remote 
                    processRoom(neighborRemote, color);
                }
            } catch (CouldNotPerformException ex) {
                throw new CouldNotPerformException("Could not process room of " + locationRemote, ex);
            }
        }
    }
}

File: module/app/preset/src/main/java/org/openbase/bco/app/preset/PartyLightTileFollowerApp.kt
package org.openbase.bco.app.preset

import org.openbase.bco.dal.control.layer.unit.app.AbstractAppController
import org.openbase.bco.dal.lib.layer.unit.Unit
import org.openbase.bco.dal.remote.action.RemoteAction
import org.openbase.bco.dal.remote.layer.unit.Units
import org.openbase.bco.dal.remote.layer.unit.location.LocationRemote
import org.openbase.bco.registry.remote.Registries
import org.openbase.jul.exception.CouldNotPerformException
import org.openbase.jul.exception.ExceptionProcessor.interruptOnShutdown
import org.openbase.jul.exception.ExceptionProcessor.isCausedBySystemShutdown
import org.openbase.jul.exception.InstantiationException
import org.openbase.jul.exception.printer.ExceptionPrinter
import org.openbase.jul.schedule.SyncObject
import org.openbase.type.domotic.action.ActionDescriptionType
import org.openbase.type.domotic.state.ActivationStateType
import org.openbase.type.domotic.unit.UnitTemplateType
import org.openbase.type.domotic.unit.location.LocationConfigType
import org.openbase.type.vision.HSBColorType
import java.util.concurrent.*

/**
 * UnitConfig
 */
class PartyLightTileFollowerApp : AbstractAppController() {
    private var actionLocationMap: MutableMap<Unit<*>?, RemoteAction>? = null
    private var locationRemoteMap: MutableMap<String, LocationRemote>? = null

    private val taskLock = SyncObject("TaskLock")

    override fun shutdown() {
        locationRemoteMap!!.clear()
        super.shutdown()
    }

    private val brightness = 0.50
    private val colors = arrayOf(
        HSBColorType.HSBColor.newBuilder().setHue(0.0).setSaturation(1.0).setBrightness(brightness).build(),
        HSBColorType.HSBColor.newBuilder().setHue(290.0).setSaturation(1.0).setBrightness(brightness).build(),
        HSBColorType.HSBColor.newBuilder().setHue(30.0).setSaturation(1.0).setBrightness(brightness).build(),
    )

    init {
        try {
            Registries.waitForData()
            this.actionLocationMap = HashMap()
            this.locationRemoteMap = HashMap()
        } catch (ex: CouldNotPerformException) {
            throw InstantiationException(this, ex)
        }
    }

    @Throws(CouldNotPerformException::class, InterruptedException::class)
    override fun execute(activationState: ActivationStateType.ActivationState): ActionDescriptionType.ActionDescription? {
        logger.debug("Execute PartyLightTileFollowerApp[$label]")

        // init tile remotes
        locationRemoteMap!!.clear()
        for (locationUnitConfig in Registries.getUnitRegistry(true)
            .getUnitConfigsByUnitType(UnitTemplateType.UnitTemplate.UnitType.LOCATION)) {
            if (locationUnitConfig.locationConfig.locationType != LocationConfigType.LocationConfig.LocationType.TILE) {
                continue
            }
            locationRemoteMap!![locationUnitConfig.id] = Units.getUnit(locationUnitConfig, false, Units.LOCATION)
        }

        TileFollower().call()
        return activationState.responsibleAction
    }

    @Throws(InterruptedException::class, CouldNotPerformException::class)
    override fun stop(activationState: ActivationStateType.ActivationState) {
        val cancelTaskList = ArrayList<Future<ActionDescriptionType.ActionDescription>>()
        for ((_, value) in actionLocationMap!!) {
            cancelTaskList.add(value.cancel())
        }

        for (cancelTask in cancelTaskList) {
            try {
                cancelTask[10, TimeUnit.SECONDS]
            } catch (ex: ExecutionException) {
                if (!isCausedBySystemShutdown(ex)) {
                    ExceptionPrinter.printHistory<Exception>("Could not cancel action!", ex, logger)
                }
            } catch (ex: TimeoutException) {
                if (!isCausedBySystemShutdown(ex)) {
                    ExceptionPrinter.printHistory<Exception>("Could not cancel action!", ex, logger)
                }
            }
        }
        super.stop(activationState)
    }

    inner class TileFollower : Callable<Void?> {
        private val processedLocations: MutableList<String> = ArrayList()

        @Throws(CouldNotPerformException::class, InterruptedException::class)
        override fun call(): Void? {
            if (locationRemoteMap!!.isEmpty()) {
                throw CouldNotPerformException("No Locations found!")
            }

            var locationRemote: LocationRemote?

            var colorIndex = 0
            while (!Thread.currentThread().isInterrupted) {
                Thread.yield()
                try {
                    // apply updates for next iteration
                    colorIndex = ++colorIndex % colors.size
                    processedLocations.clear()

                    // select initial room
                    locationRemote = locationRemoteMap!![config.placementConfig.locationId]

                    processRoom(locationRemote, colors[colorIndex])
                } catch (ex: CouldNotPerformException) {
                    ExceptionPrinter.printHistory(
                        CouldNotPerformException(
                            "Skip animation run!",
                            interruptOnShutdown(ex)
                        ), logger
                    )
                }
            }
            return null
        }

        @Throws(CouldNotPerformException::class, InterruptedException::class)
        private fun processRoom(locationRemote: LocationRemote?, color: HSBColorType.HSBColor) {
            logger.debug("Set $locationRemote to $color...")
            try {
                // skip if no colorable light is present

                if (Registries.getUnitRegistry().getUnitConfigsByLocationIdAndUnitType(
                        locationRemote!!.id,
                        UnitTemplateType.UnitTemplate.UnitType.COLORABLE_LIGHT
                    ).isNotEmpty()
                ) {
                    if (locationRemote.isConnected && locationRemote.isDataAvailable) {
                        val remoteAction = observe(locationRemote.setColor(color, defaultActionParameter))
                        actionLocationMap!![locationRemote] = remoteAction
                        remoteAction.waitForRegistration()
                    }
                    Thread.sleep(1000)
                }

                // mark as processed
                processedLocations.add(locationRemote.id)

                // process neighbors
                var neighborRemote: LocationRemote?
                for (neighborConfig in Registries.getUnitRegistry().getNeighborLocationsByLocationId(
                    locationRemote.id
                )) {
                    // skip if already processed
                    if (processedLocations.contains(neighborConfig.id)) {
                        continue
                    }

                    neighborRemote = locationRemoteMap!![neighborConfig.id]

                    // process remote 
                    processRoom(neighborRemote, color)
                }
            } catch (ex: CouldNotPerformException) {
                throw CouldNotPerformException("Could not process room of $locationRemote", ex)
            }
        }
    }
}
